libp2p/protocols/pubsub/pubsub.nim

```diff
@@ -202,26 +198,9 @@ proc sendSubs*(p: PubSub,
       else:
         libp2p_pubsub_broadcast_unsubscriptions.inc(labelValues = ["generic"])
 
-method subscribeTopic*(p: PubSub,
-                       topic: string,
-                       subscribe: bool,
-                       peer: PubSubPeer) {.base.} =
-  # both gossipsub and floodsub diverge, and this super call is not necessary right now
-  # if necessary remove the assertion
-  doAssert(false, "unexpected call to pubsub.subscribeTopic")
-
-method rpcHandler*(p: PubSub,
-                   peer: PubSubPeer,
-                   rpcMsg: RPCMsg): Future[void] {.base.} =
-  ## handle rpc messages
-  trace "processing RPC message", msg = rpcMsg.shortLog, peer
-  for i in 0..<min(rpcMsg.subscriptions.len, p.topicsHigh):
-    let s = rpcMsg.subscriptions[i]
-    trace "about to subscribe to topic", topicId = s.topic, peer, subscribe = s.subscribe
-    p.subscribeTopic(s.topic, s.subscribe, peer)
-
+proc updateMetrics*(p: PubSub, rpcMsg: RPCMsg) =
   for i in 0..<min(rpcMsg.subscriptions.len, p.topicsHigh):
-    let sub = rpcMsg.subscriptions[i]
+    template sub(): untyped = rpcMsg.subscriptions[i]
```

Comment:
I really dislike this because it's using a clever and unintuitive trick to create an alias. It's also masking away potential rase conditions, if this proc becomes async it would be very easy to miss why we suddenly get NPEs or out of bounds access. I would rather we use `var sub = addr rpcMsg.subscriptions[i] # do something with sub[]` which makes things explicit and hard to miss.

Reply:
I think you might have missed something here, `addr` is the construct that disables the safety mechanisms in Nim and permits the kind of issues you're referring to because the resulting pointer can become dangling - it should be avoided when possible.
also, it would not be applicable here since this is not a mutable instance.

template is used throughout the codebase to work around nim 1.2 issues in performance hotspots. gossipsub is the biggest memory consumer in nimbus and this is one of the sources of that consumption.

Reply:
> addr is the construct that disables the safety mechanisms in Nim

Nope, didn't miss that. What I'm referring to here is explicitness. The template in this case is "pretending" that the element has been placed in a variable - for all intents and purposes it's usage looks and feels like that, like here for example - https://github.com/status-im/nim-libp2p/pull/569/files#diff-a99c4291e04a9bde36fc1088e4d03053d57e0fcb54d659fcbc4b313362cf7278R206, this is misleading. 

I don't think that the compiler can deduce the size of the array/sequence in this case, so there are no compile checks for out of bounds access and you effectively getting similar guarantees to `addr`.

The problem with masking it like this is that someone can innadverdly do something like:

```nim
template sub(): untyped = rpcMsg.subscriptions[i]
....
del rpcMsg.subscriptions[i]
...
# uhoh - in the best case you're going to crash, in the worst it's you're pointing to the wrong element
echo sub.topic
```

It won't happen now, because the proc is not `async`, but this tend to change over the lifetime of the implementation.

Reply:
> deduce the size

the `[]` operator checks for out-of-bounds at runtime. the pointer returned by `addr` is not checked any further.

> similar guarantees to addr.

`addr` returns a `ptr T` - a pointer can for example be copied into a closure, but when that is done, the data that it points to is not copied - this causes it to become dangling, the very thing you're trying to argue is a problem with the template. `addr` is one of the memory-unsafe constructs in Nim - it doesn't offer any guarantees at all. If you were to use `addr` with `async`, you would be circumventing the memory-safety that the language gives you and crash where a template like this wouldn't (it would take a copy of `rpcMsg` and keep working, albeit inefficiently)

> this is misleading.

Generally, `rpcMsg` is immutable meaning that it cannot change the way you're suggesting in your snippet - you also cannot use `addr` on it, for similar reasons: you would now be able to mutate it and cause these kinds of issues.

In cases like this where the source data is immutable, in Nim 1.2 you can either take a copy with `let` (what the previous code did) or use a template - the language doesn't allow you to use `addr` for good reasons, and even if it did, it would cause worse problems if the function was refactored or turned into async, as noted above.

We can forgo this optimization of course, but doing so comes at a measurable cost.